pkg/metrics/metrics_test.go

```diff
@@ -315,38 +272,209 @@ func TestWriteMetric(t *testing.T) {
 		t.Run(tc.name, func(t *testing.T) {
 			t.Parallel()
 
+			var saveReq *http.Request
+
+			// Request body is intentionally leaked to allow for inspection in test cases.
+			ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				if saveReq != nil {
+					t.Fatal("multiple requests in a single test")
+				}
+				saveReq = r.Clone(context.Background())
+				body, err := io.ReadAll(r.Body)
+				if err != nil {
+					t.Fatalf("error copying body: %s", err.Error())
+				}
+				saveReq.Body = io.NopCloser(bytes.NewBuffer(body))
+				r.Body = io.NopCloser(bytes.NewBuffer(body))
+
+				if !strings.HasSuffix(r.RequestURI, "/sendMetrics") {
+					w.WriteHeader(http.StatusNotFound)
+					fmt.Fprintln(w, http.StatusText(http.StatusNotFound))
+					return
+				}
+
+				if strings.HasSuffix(r.RequestURI, "400/sendMetrics") {
+					w.WriteHeader(http.StatusBadRequest)
+					fmt.Fprintln(w, "bad request")
+					return
+				}
+
+				if strings.HasSuffix(r.RequestURI, "500/sendMetrics") {
+					w.WriteHeader(http.StatusInternalServerError)
+					fmt.Fprintln(w, "internal error")
+					return
+				}
+
+				w.WriteHeader(http.StatusOK)
+				fmt.Fprintln(w, "ok")
+			}))
+
+			t.Cleanup(ts.Close)
+
 			ctx := context.Background()
-			relativePath := fmt.Sprintf("/%d", rand.Uint64()) //nolint:gosec
+			var relativePath string
 			if tc.responseCodeOverride != 0 {
-				relativePath = fmt.Sprintf("%s/%d", relativePath, tc.responseCodeOverride)
+				relativePath = fmt.Sprintf("%d", tc.responseCodeOverride)
 			}
-			tc.client.Config.ServerURL = fmt.Sprintf("%s%s", ts.URL, relativePath)
+			tc.client.Config.ServerURL = fmt.Sprintf("%s/%s", ts.URL, relativePath)
```

Comment:
This is all very weird now. You have one server per test, so there's no need for the indirection of the URL:

```go
ts := httptest.NewTestServer(http.HandlerFunc(w http.ResponseWriter, r *http.Request) {
  // move validation logic about the request body in here

  w.WriteHeader(tc.responseCode)
})
t.Cleanup(ts.Close)
```